cluster_tools/cluster_tools/schedulers/cluster_executor.py

```diff
@@ -158,26 +141,68 @@ def as_completed(cls, futs: list[Future[_T]]) -> Iterator[Future[_T]]:
     def executor_key(cls) -> str:
         pass
 
-    def handle_kill(
-        self, existing_sigint_handler: Any, signum: int | None, frame: Any
+    @classmethod
+    def _install_signal_handlers(cls) -> None:
+        # Only overwrite the signal handler once
+        if cls._installed_signal_handler:
+            return
+
+        # Clean up if a SIGINT or SIGTERM signal is received. However, do not
+        # interfere with the existing signal handler of the process and execute
+        # it afterwards.
+        existing_sigint_handler = signal.getsignal(signal.SIGINT)
+        signal.signal(
+            signal.SIGINT,
+            partial(cls._handle_shutdown, existing_sigint_handler),
+        )
+        existing_sigterm_handler = signal.getsignal(signal.SIGTERM)
+        signal.signal(
+            signal.SIGTERM,
+            partial(cls._handle_shutdown, existing_sigterm_handler),
+        )
+
+        cls._installed_signal_handler = True
+
+    @classmethod
+    def _register_shutdown_hook(cls, hook: Callable[[], None]) -> None:
+        cls._shutdown_hooks.append(hook)
```

Comment:
`_register_shutdown_hook` is only called in `__init__`. why is it a list then? couldnt it be an `Optional[Callable]` and `_installed_signal_handler` could also be absorbed, because either the hook was attached (is not none) or not.
additionally, I think that one could merge `_register_shutdown_hook` and `_install_signal_handlers` into one function, because one simply delegates to the other (plus appending to a list). I feel like this would be a bit easier to understand then.